src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Internal/ValidationStack.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Mvc.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Internal
+{
+    internal class ValidationStack
+    {
+        // We tested the performance of a list at size 15 and found it still better than hashset, but to avoid a costly
+        // O(n) search at larger n we set the cutoff to 20. If someone finds the point where they intersect feel free to change this number.
+        protected virtual int CutOff
```

Comment:
For @pranavkm points and consistency, `protected virtual int CutOff => 20;`. And, should this property be `static` and / or `internal` since it's apparently exposed only for tests (we don't need to subclass it in product code)?

Also, this is not currently used in tests. Suggest changing tests containing magic numbers like `23` to use this.

`internal const int CutOff = 20;` probably the easiest to use since this could be useful in `[InlineData]`.